cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewPluginFinder.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.preview;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.common.ArtifactNotFoundException;
+import io.cdap.cdap.common.internal.remote.RemoteClientFactory;
+import io.cdap.cdap.common.io.Locations;
+import io.cdap.cdap.internal.app.runtime.artifact.RemotePluginFinder;
+import io.cdap.cdap.internal.app.worker.sidecar.ArtifactLocalizerClient;
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+
+import java.io.IOException;
+
+/**
+ * PreviewPluginFinder is an extension of {@link RemotePluginFinder} that is meant to be used exclusively in tasks
+ * running in the {@link PreviewRunnerTwillRunnable}. This implementation uses the {@link ArtifactLocalizerClient} to
+ * download and cache the given artifact on the local file system; however, it does not unpack the artifact as the
+ * SparkCompiler may require it packed to function properly. See CDAP-19311 for details.
```

Comment:
CDAP-19311 doesn't have much detail in it. Can you add more detail, like the full stack trace?

Do we know the performance impact of making this change? I guess there isn't a good way to only avoid unpacking for things that use the SparkCompiler?

Reply:
Added the two stacktraces to the ticket.

Performance implication should not be significant. The only situation in which this would be an optimization would be if pipelines were running with the same artifacts back to back saturating the preview runners; however, the real bottleneck should be the transfer of the artifacts themselves and not the unpacking.

I did some quick investigation and I'm not sure if we can force the scala IMain compiler to use a java classpath with an unpacked jar file. We don't have much of way to avoid unpacking things for the spark compiler unless it supports some form of reading classes from a directory

Reply:
yeah that's fine, though it would be good to put that in the ticket as well.